dali/operators/reader/loader/coco_loader.h

```diff
@@ -24,53 +24,127 @@
+  return spec.HasArgument("save_preprocessed_annotations_dir") ||
+    (spec.HasArgument("dump_meta_files_path") && spec.GetArgument<bool>("dump_meta_files_path"));
+}
+
 class DLL_PUBLIC CocoLoader : public FileLabelLoader {
  public:
-  explicit inline CocoLoader(
-    const OpSpec& spec,
-    std::vector<int> &heights,
-    std::vector<int> &widths,
-    std::vector<int> &offsets,
-    std::vector<float> &boxes,
-    std::vector<int> &labels,
-    std::vector<int> &counts,
-    std::vector<std::vector<int> > &masks_meta,
-    std::vector<std::vector<float> > &masks_coords,
-    std::vector<std::vector<std::string> > &masks_rles,
-    std::vector<std::vector<int> > &masks_rles_idx,
-    bool read_masks,
-    bool pixelwise_masks,
-    bool save_img_ids,
-    std::vector<int> &original_ids,
-    bool shuffle_after_epoch = false) :
-      FileLabelLoader(spec, std::vector<std::pair<string, int>>(), shuffle_after_epoch),
-      spec_(spec),
-      parse_meta_files_(spec.HasArgument("meta_files_path")),
-      heights_(heights),
-      widths_(widths),
-      offsets_(offsets),
-      boxes_(boxes),
-      labels_(labels),
-      counts_(counts),
-      masks_meta_(masks_meta),
-      masks_coords_(masks_coords),
-      masks_rles_(masks_rles),
-      masks_rles_idx_(masks_rles_idx),
-      read_masks_(read_masks),
-      pixelwise_masks_(pixelwise_masks),
-      save_img_ids_(save_img_ids),
-      original_ids_(original_ids) {}
+  explicit inline CocoLoader(const OpSpec& spec)
+      : FileLabelLoader(spec, std::vector<std::pair<string, int>>(),
+                        spec.GetArgument<bool>("shuffle_after_epoch")),
+        spec_(spec) {
+    has_preprocessed_annotations_ = HasPreprocessedAnnotations(spec);
+    DALI_ENFORCE(has_preprocessed_annotations_ || spec.HasArgument("annotations_file"),
+        "Either ``annotations_file`` or ``preprocessed_annotations`` must be provided");
+    if (has_preprocessed_annotations_) {
+      for (const char* arg_name : {"annotations_file", "skip_empty", "ratio", "ltrb",
+                                   "size_threshold", "dump_meta_files", "dump_meta_files_path"}) {
+        if (spec.HasArgument(arg_name))
+          DALI_FAIL(make_string("When reading data from preprocessed annotation files, \"",
+                                arg_name, "\" is not supported."));
+      }
+    }
+
+    output_polygon_masks_ = OutPolygonMasksEnabled(spec);
+    output_pixelwise_masks_ = OutPixelwiseMasksEnabled(spec);
+    output_image_ids_ = OutImageIdsEnabled(spec);
+    if (output_polygon_masks_ && output_pixelwise_masks_) {
+      DALI_FAIL("``pixelwise_masks`` and ``polygon_masks`` are mutually exclusive");
+    }
+
+    if (HasSavePreprocessedAnnotations(spec) != HasSavePreprocessedAnnotationsDir(spec)) {
+      DALI_FAIL("``save_preprocessed_annotations`` and ``save_preprocessed_annotations_dir`` "
+                "should be provided together");
+    }
+  }
+
+  struct BboxesInfo {
```

Comment:
not used, to be removed

Reply:
So why not to remove now?

Reply:
Just haven't pushed changes yet, don't worry I'll remove